### PR TITLE
[ios][notifications] fix: prevent initial notification being swallowed if app is closed

### PR DESCRIPTION
**Pull request by cruzach**
# What is the issue?

`Notifications.addResponseReceivedListener` allows users to handle interactions with notifications that come in while the app is foregrounded, backgrounded, and killed. In the bare workflow (on iOS at least), this listener works perfectly as is. However, in the managed workflow, the listener is never fired when tapping a notification **while the app is killed** (listener works if you tap a notification while the app is open or backgrounded). 

See https://github.com/expo/expo/issues/6943 for more info & reproducible demo. This PR addresses the iOS side of https://github.com/expo/expo/issues/6943

# What is the root cause?

The event emitter behind `Notifications.addResponseReceivedListener`  is not listening when the app first opens, so `expo-notifications` _should_ store the notification in [`_pendingNotificationResponses`](https://github.com/expo/expo/blob/master/packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationCenterDelegate.m#L136), and then [handle it here](https://github.com/expo/expo/blob/master/packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationCenterDelegate.m#L184-L196) once the delegate is added later on. This works like a charm in the bare workflow because `responseWillBeHandledByAnyDelegate` remains false. 

**But** in the managed workflow, [`EXUserNotificationManager` is added as a delegate of `EXNotificationCenterDelegate`](https://github.com/expo/expo/blob/master/ios/Exponent/ExpoKit/ExpoKit.m#L121) and makes the latter believe that the notification response will be handled by EXUserNotificationManager, and thus never adds the response to `_pendingNotificationResponses` to be handled later when `EXNotificationsEmitter` is added as a delegate.

# How

Fixed by adding a pendingNotificationResponse by default, then going on to also send the response to a delegate (if there is one). This way both the legacy module calls its `didReceiveNotificationResponse` method, **and** `EXNotificationsEmitter` calls it's own once the delegate is added on startup.

This doesn't seem to have any adverse effects in any of the situations I tested below, but maybe there's an edge case I'm not considering that the code I'm suggesting to remove handles? If that's the case I can revisit this from another angle

# Test Plan

Tested in the bare workflow, Expo Client, and Expokit (closest i could get to standalone app with these changes) using the code in this snack to show both legacy notifications and `expo-notifications` triggers https://snack.expo.io/@charliecruzan/notificationsresponsetrigger

- tap notification while app foregrounded, `LegacyNotifications.addListener` and `Notifications.addNotificationResponseReceivedListener` each triggered once ✅ 
- tap notification while app backgrounded, `LegacyNotifications.addListener` and `Notifications.addNotificationResponseReceivedListener` each triggered once ✅ 
- tap notification after killing app, `LegacyNotifications.addListener` and `Notifications.addNotificationResponseReceivedListener` each triggered once ✅

### CODE REVIEW

**Comment by macuna** [comment]
Did you tested this in Android @cruzach ?

I tested the code provided but still I can't handle the response in the notification even with legacy notification when the app is killed.

**Comment by cruzach** [comment]
@macuna this addresses iOS only and I'm not sure what you mean that you tested the code. This fix applies to the managed workflow only, and includes native code changes

**Review by lukmccall — COMMENTED** [review]
I'm not sure if this doesn't break anything else. Maybe, it will be better to leave the old logic, but try to handle it differently. Maybe by checking if `EXUserNotificationManager` is the only receiver? Hovewer, I'm not sure if it possible here. What do you think? @cruzach

**Comment by cruzach** [comment]
Yeah I had considered that, the drawback is that would leak code that _only_ applies to the managed workflow into the `expo-notifications` module. That being said, it would only need to remain for as long as we support the legacy notifications module, so I guess it's not too big of a problem

**Comment by cruzach** [comment]
It is possible, just need to compare class by a string literal so that we don't need to import `Expokit` to `EXNotifications`, eg:
```
[NSStringFromClass([pointer class])  isEqual: @"EXUserNotificationManager"]
```

File: packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationCenterDelegate.m [hunk 1]
```diff
@@ -124,18 +124,18 @@ __block void (^completionHandlerCaller)(void) = ^{
...
+      // Remove EXUserNotificationManager check when LegacyNotifications are no longer supported
+      responseWillBeHandledByAppropriateDelegate = YES;
       break;
     }
   }
-  if (!responseWillBeHandledByAnyDelegate) {
+  if (!responseWillBeHandledByAppropriateDelegate) {
     [_pendingNotificationResponses addObject:response];
-    completionHandler();
-    return;
```
Comment (cruzach) [inline packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationCenterDelegate.m:138]: completion handler will be called [later on](https://github.com/expo/expo/blob/master/packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationCenterDelegate.m#L166), so it's okay to remove it from here. And we don't want to `return` here, since `EXUserNotificationManager` may still need to handle the notification response

**Review by mczernek — APPROVED** [review]

File: packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationCenterDelegate.m [hunk 2]
```diff
@@ -124,18 +124,18 @@ __block void (^completionHandlerCaller)(void) = ^{
...
 {
   // Save response to pending responses array if none of the handlers will handle it.
-  BOOL responseWillBeHandledByAnyDelegate = NO;
+  BOOL responseWillBeHandledByAppropriateDelegate = NO;
   for (int i = 0; i < _delegates.count; i++) {
     id pointer = [_delegates pointerAtIndex:i];
-    if ([pointer respondsToSelector:@selector(userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:)]) {
-      responseWillBeHandledByAnyDelegate = YES;
+    if ([pointer respondsToSelector:@selector(userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:)] 
+      && ![NSStringFromClass([pointer class])  isEqual: @"EXUserNotificationManager"]) {
```
Comment (lukmccall) [inline packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationCenterDelegate.m:131]:
```suggestion
      && ![NSStringFromClass([pointer class]) isEqual: @"EXUserNotificationManager"]) {
```

**Review by lukmccall — APPROVED** [review]
LGTM 💪💪💪

**Comment by summerkiflain** [comment]
@cruzach Sorry If I am being naive, but when will be this fix available for managed workflows? can I rebuild the ipa to get this fixed, or do we have to wait for the next sdk release to get this? I would like to revert back to legacy notifications if latter is the case.

**Comment by sjchmiela** [comment]
I don't think we're planning another update to SDK38 iOS builders or Expo client, so I think the latter is true.

**Comment by summerkiflain** [comment]
@sjchmiela thanks for replying, I was already planning to revert back to legacy notifications API due to https://github.com/expo/expo/issues/9382, I would have never moved to `"useNextNotificationsApi": true` if the old notification api was not renamed to `LegacyNotifications`.

**Comment by sjchmiela** [comment]
Truth be told, it's not easy to test a module on all devices with all different configurations, so hitting that one device configuration for which the module is crashing in #9382 wouldn't have been easy. 🙂 

`LegacyNotifications` are legacy for a reason, they're weren't bare-workflow-compatible and since we're moving all unimodules to be bare-workflow-compatible and we don't have the bandwidth to support two different notifications implementations one will need to be eventually abandoned. 😕 Knowing that there may be some unexpected bugs we let you switch between the APIs so easily — we'll only remove `LegacyNotifications` once we're sure that users can move to the new API without problems.

On the other hand note that `useNextNotificationsApi` is only used on Android — on iOS both APIs work alongside each other and you don't need to rebuild IPA to be able to move back to the legacy one.

**Comment by sjchmiela** [comment]
BTW we also highly appreciate all the developers who have the courage to move to `expo-notifications` — while the API is stable, we've had some unexpected problems with how stuff was implemented (eg. undocumented Android restrictions) — each issue report will get the library closer and closer to being 100%-safe. 🎖️

**Comment by summerkiflain** [comment]
@sjchmiela I totally get your points, we really appreciate everyone working on the new notification api to stabilize it for the future, but we just got confused with the naming of old api although It was pretty stable and the new api not being able to redirect users to the desired screen on the tap actions is the biggest factor new api let down the users, a lot of our users are having this problem, the thing is due to naming it legacyNotifications we move both our prod and beta apps to new API, I would have preferred to switch to new API only on my beta app and keep my prod app on the same api which was working fine already at least to user actions on push notifications, I had to add scheduled notifications to the app and I am not sure I missed or legacyNotification docs became available a little later, I could not find any other option but to move to new API.
still its a bit confusing for android, 1 api is legacy and 1 is next, so no current stable version to stick to for prod apps, users should have been given at least 1 sdk margin before naming the old one to Legacy. and new API is not stable at least not yet for me and many other users.

**Comment by summerkiflain** [comment]
@cruzach @sjchmiela I have migrated my app to expo sdk 39, followed the migrations instruction properly + the instructions mentioned in sdk39++ thread to get the fixes for sdk39, generated new ipa to test this, but the response listener is still not triggering for me on iOS, I have a class based main app container and I am registering listener in it's constructor, which always works in android, but never works in ios when app is killed, works fine when app is in background or foreground.
any idea what I should be checking first for proper solution?
